File: soc/bcm2710/soc_init.c
/*
 * soc/bcm2710/soc_init.c - BCM2710 Platform Initialization
 *
 * Tutorial-OS: BCM2710 HAL Implementation
 *
 * Implements hal_platform.h for BCM2710/BCM2837.
 * This is the entry point for all BCM2710-specific initialization.
 */

#include "hal/hal_platform.h"
#include "hal/hal_timer.h"
#include "hal/hal_gpio.h"
#include "bcm2710_regs.h"
#include "bcm2710_mailbox.h"

/* =============================================================================
 * INTERNAL STATE
 * =============================================================================
 */

static bool g_platform_initialized = false;

/* Board name based on revision code */
static const char *g_board_name = "Raspberry Pi (BCM2710)";
static const char *g_soc_name = "BCM2710";
static uint32_t g_board_revision = 0;
static uint64_t g_board_serial = 0;

/* =============================================================================
 * BOARD IDENTIFICATION
 * =============================================================================
 * Decode Raspberry Pi board revision codes.
 */

static const char *decode_board_name(uint32_t revision)
{
    /*
     * New-style revision codes (bit 23 set):
     * Bits 4-11: Type
     */
    if (revision & (1 << 23)) {
        uint32_t type = (revision >> 4) & 0xFF;
        switch (type) {
            case 0x04: return "Raspberry Pi 2B";
            case 0x08: return "Raspberry Pi 3B";
            case 0x0D: return "Raspberry Pi 3B+";
            case 0x0E: return "Raspberry Pi 3A+";
            case 0x10: return "Raspberry Pi CM3+";
            case 0x12: return "Raspberry Pi Zero 2W";
            default:   return "Raspberry Pi (BCM2710)";
        }
    }

    /* Old-style revision - just return generic */
    return "Raspberry Pi";
}

/* =============================================================================
 * INITIALIZATION
 * =============================================================================
 */

hal_error_t hal_platform_early_init(void)
{
    /*
     * Early init - called before most other subsystems.
     * On BCM2710, there's not much to do here.
     */
    return HAL_SUCCESS;
}

hal_error_t hal_platform_init(void)
{
    hal_error_t err;

    if (g_platform_initialized) {
        return HAL_ERROR_ALREADY_INIT;
    }

    /* Initialize timer subsystem */
    err = hal_timer_init();
    if (HAL_FAILED(err)) {
        return err;
    }

    /* Initialize GPIO subsystem */
    err = hal_gpio_init();
    if (HAL_FAILED(err)) {
        return err;
    }

    /* Query board information via mailbox */
    bcm_mailbox_get_board_revision(&g_board_revision);
    bcm_mailbox_get_board_serial(&g_board_serial);
    g_board_name = decode_board_name(g_board_revision);

    g_platform_initialized = true;
    return HAL_SUCCESS;
}

bool hal_platform_is_initialized(void)
{
    return g_platform_initialized;
}

/* =============================================================================
 * PLATFORM INFORMATION
 * =============================================================================
 */

hal_error_t hal_platform_get_info(hal_platform_info_t *info)
{
    if (info == NULL) {
        return HAL_ERROR_NULL_PTR;
    }

    info->platform_id = HAL_PLATFORM_RPI_ZERO2W;  /* Default to Zero 2W */
    info->arch = HAL_ARCH_ARM64;
    info->board_name = g_board_name;
    info->soc_name = g_soc_name;
    info->board_revision = g_board_revision;
    info->serial_number = g_board_serial;

    return HAL_SUCCESS;
}

hal_platform_id_t hal_platform_get_id(void)
{
    return HAL_PLATFORM_RPI_ZERO2W;
}

const char *hal_platform_get_board_name(void)
{
    return g_board_name;
}

const char *hal_platform_get_soc_name(void)
{
    return g_soc_name;
}

/* =============================================================================
 * MEMORY INFORMATION
 * =============================================================================
 */

hal_error_t hal_platform_get_memory_info(hal_memory_info_t *info)
{
    if (info == NULL) {
        return HAL_ERROR_NULL_PTR;
    }

    uint32_t arm_base, arm_size, vc_base, vc_size;

    if (!bcm_mailbox_get_arm_memory(&arm_base, &arm_size)) {
        return HAL_ERROR_HARDWARE;
    }

    if (!bcm_mailbox_get_vc_memory(&vc_base, &vc_size)) {
        /* VC memory query is optional */
        vc_base = 0;
        vc_size = 0;
    }

    info->arm_base = arm_base;
    info->arm_size = arm_size;
    info->gpu_base = vc_base;
    info->gpu_size = vc_size;
    info->peripheral_base = BCM_PERIPHERAL_BASE;

    return HAL_SUCCESS;
}

size_t hal_platform_get_arm_memory(void)
{
    uint32_t base, size;
    if (bcm_mailbox_get_arm_memory(&base, &size)) {
        return size;
    }
    return 0;
}

size_t hal_platform_get_total_memory(void)
{
    uint32_t arm_base, arm_size, vc_base, vc_size;

    if (!bcm_mailbox_get_arm_memory(&arm_base, &arm_size)) {
        return 0;
    }

    if (bcm_mailbox_get_vc_memory(&vc_base, &vc_size)) {
        return arm_size + vc_size;
    }

    return arm_size;
}

/* =============================================================================
 * CLOCK INFORMATION
 * =============================================================================
 */

hal_error_t hal_platform_get_clock_info(hal_clock_info_t *info)
{
    if (info == NULL) {
        return HAL_ERROR_NULL_PTR;
    }

    bcm_mailbox_get_clock_rate(BCM_CLOCK_ARM, &info->arm_freq_hz);
    bcm_mailbox_get_clock_rate(BCM_CLOCK_CORE, &info->core_freq_hz);
    bcm_mailbox_get_clock_rate(BCM_CLOCK_UART, &info->uart_freq_hz);
    bcm_mailbox_get_clock_rate(BCM_CLOCK_EMMC, &info->emmc_freq_hz);

    return HAL_SUCCESS;
}

uint32_t hal_platform_get_arm_freq(void)
{
    uint32_t freq;
    if (bcm_mailbox_get_clock_rate(BCM_CLOCK_ARM, &freq)) {
        return freq;
    }
    return 0;
}

uint32_t hal_platform_get_arm_freq_measured(void)
{
    uint32_t freq;
    if (bcm_mailbox_get_clock_measured(BCM_CLOCK_ARM, &freq)) {
        return freq;
    }
    return 0;
}

uint32_t hal_platform_get_clock_rate(hal_clock_id_t clock_id)
{
    uint32_t bcm_clock;

    switch (clock_id) {
        case HAL_CLOCK_ARM:     bcm_clock = BCM_CLOCK_ARM; break;
        case HAL_CLOCK_CORE:    bcm_clock = BCM_CLOCK_CORE; break;
        case HAL_CLOCK_UART:    bcm_clock = BCM_CLOCK_UART; break;
        case HAL_CLOCK_EMMC:    bcm_clock = BCM_CLOCK_EMMC; break;
        case HAL_CLOCK_PWM:     bcm_clock = BCM_CLOCK_PWM; break;
        default:                return 0;
    }

    uint32_t freq;
    if (bcm_mailbox_get_clock_rate(bcm_clock, &freq)) {
        return freq;
    }
    return 0;
}

/* =============================================================================
 * TEMPERATURE
 * =============================================================================
 */

hal_error_t hal_platform_get_temperature(int32_t *temp_mc)
{
    if (temp_mc == NULL) {
        return HAL_ERROR_NULL_PTR;
    }

    uint32_t temp;
    if (!bcm_mailbox_get_temperature(&temp)) {
        return HAL_ERROR_HARDWARE;
    }

    *temp_mc = (int32_t)temp;
    return HAL_SUCCESS;
}

int32_t hal_platform_get_temp_celsius(void)
{
    uint32_t temp_mc;
    if (bcm_mailbox_get_temperature(&temp_mc)) {
        return (int32_t)(temp_mc / 1000);
    }
    return -1;
}

hal_error_t hal_platform_get_max_temperature(int32_t *temp_mc)
{
    if (temp_mc == NULL) {
        return HAL_ERROR_NULL_PTR;
    }

    uint32_t temp;
    if (!bcm_mailbox_get_max_temperature(&temp)) {
        return HAL_ERROR_HARDWARE;
    }

    *temp_mc = (int32_t)temp;
    return HAL_SUCCESS;
}

/* =============================================================================
 * THROTTLE STATUS
 * =============================================================================
 */

hal_error_t hal_platform_get_throttle_status(uint32_t *status)
{
    if (status == NULL) {
        return HAL_ERROR_NULL_PTR;
    }

    if (!bcm_mailbox_get_throttled(status)) {
        return HAL_ERROR_HARDWARE;
    }

    return HAL_SUCCESS;
}

bool hal_platform_is_throttled(void)
{
    uint32_t status;
    if (bcm_mailbox_get_throttled(&status)) {
        /* Check if currently throttled (bits 0-3) */
        return (status & 0xF) != 0;
    }
    return false;
}

/* =============================================================================
 * POWER MANAGEMENT
 * =============================================================================
 */

hal_error_t hal_platform_set_power(hal_device_id_t device, bool on)
{
    uint32_t bcm_device;

    switch (device) {
        case HAL_DEVICE_SD_CARD:    bcm_device = BCM_DEVICE_SD; break;
        case HAL_DEVICE_UART0:      bcm_device = BCM_DEVICE_UART0; break;
        case HAL_DEVICE_UART1:      bcm_device = BCM_DEVICE_UART1; break;
        case HAL_DEVICE_USB:        bcm_device = BCM_DEVICE_USB; break;
        default:                    return HAL_ERROR_INVALID_ARG;
    }

    if (!bcm_mailbox_set_power_state(bcm_device, on)) {
        return HAL_ERROR_HARDWARE;
    }

    return HAL_SUCCESS;
}

hal_error_t hal_platform_get_power(hal_device_id_t device, bool *on)
{
    if (on == NULL) {
        return HAL_ERROR_NULL_PTR;
    }

    uint32_t bcm_device;

    switch (device) {
        case HAL_DEVICE_SD_CARD:    bcm_device = BCM_DEVICE_SD; break;
        case HAL_DEVICE_UART0:      bcm_device = BCM_DEVICE_UART0; break;
        case HAL_DEVICE_UART1:      bcm_device = BCM_DEVICE_UART1; break;
        case HAL_DEVICE_USB:        bcm_device = BCM_DEVICE_USB; break;
        default:                    return HAL_ERROR_INVALID_ARG;
    }

    if (!bcm_mailbox_get_power_state(bcm_device, on)) {
        return HAL_ERROR_HARDWARE;
    }

    return HAL_SUCCESS;
}

/* =============================================================================
 * SYSTEM CONTROL
 * =============================================================================
 */

hal_error_t hal_platform_reboot(void)
{
    /* Use watchdog to trigger reboot */
    /* TODO: Implement watchdog reset */
    return HAL_ERROR_NOT_SUPPORTED;
}

hal_error_t hal_platform_shutdown(void)
{
    /* Not directly supported on Pi without PMIC */
    return HAL_ERROR_NOT_SUPPORTED;
}

void hal_panic(const char *message)
{
    (void)message;
    /* TODO: Display panic message if possible */

    /* Disable interrupts and halt */
    while (1) {
        HAL_WFI();
    }
}

/* =============================================================================
 * DEBUG OUTPUT (Minimal implementation)
 * =============================================================================
 */

void hal_debug_putc(char c)
{
    (void)c;
    /* TODO: UART output */
}

void hal_debug_puts(const char *s)
{
    while (*s) {
        hal_debug_putc(*s++);
    }
}

void hal_debug_printf(const char *fmt, ...)
{
    (void)fmt;
    /* TODO: Implement minimal printf */
}

File: soc/bcm2710/src/soc_init.rs
//! BCM2710 platform initialization and information queries.
//!
//! Implements [`hal::platform::Platform`] for BCM2710/BCM2837.
//! All hardware queries go through the VideoCore mailbox.

use hal::platform::*;
use crate::{mailbox, regs, Bcm2710Display, Bcm2710Timer};

/// BCM2710 platform instance.
pub struct Bcm2710Platform {
    initialized: bool,
    board_name: &'static str,
    board_revision: u32,
    board_serial: u64,
}

impl Bcm2710Platform {
    pub const fn new() -> Self {
        Self {
            initialized: false,
            board_name: "Raspberry Pi (BCM2710)",
            board_revision: 0,
            board_serial: 0,
        }
    }
}

/// Decode the Raspberry Pi board name from the revision code.
fn decode_board_name(revision: u32) -> &'static str {
    // New-style revision codes have bit 23 set; bits 4–11 encode the type.
    if revision & (1 << 23) != 0 {
        match (revision >> 4) & 0xFF {
            0x04 => "Raspberry Pi 2B",
            0x08 => "Raspberry Pi 3B",
            0x0D => "Raspberry Pi 3B+",
            0x0E => "Raspberry Pi 3A+",
            0x10 => "Raspberry Pi CM3+",
            0x12 => "Raspberry Pi Zero 2W",
            _    => "Raspberry Pi (BCM2710)",
        }
    } else {
        "Raspberry Pi"
    }
}

/// Map a HAL [`ClockId`] to the BCM mailbox clock constant.
fn hal_to_bcm_clock(id: ClockId) -> Option<u32> {
    match id {
        ClockId::Arm   => Some(regs::clock::ARM),
        ClockId::Core  => Some(regs::clock::CORE),
        ClockId::Uart  => Some(regs::clock::UART),
        ClockId::Emmc  => Some(regs::clock::EMMC),
        ClockId::Pwm   => Some(regs::clock::PWM),
        ClockId::Pixel => Some(regs::clock::PIXEL),
    }
}

/// Map a HAL [`DeviceId`] to the BCM mailbox device constant.
fn hal_to_bcm_device(id: DeviceId) -> Option<u32> {
    match id {
        DeviceId::SdCard => Some(regs::device::SD),
        DeviceId::Uart0  => Some(regs::device::UART0),
        DeviceId::Uart1  => Some(regs::device::UART1),
        DeviceId::Usb    => Some(regs::device::USB),
        _                => None,
    }
}

impl Platform for Bcm2710Platform {
    type Disp = Bcm2710Display;
    type Tmr  = Bcm2710Timer;

    fn create_display(&self) -> Self::Disp { Bcm2710Display::new() }
    fn timer(&self) -> Self::Tmr { Bcm2710Timer }

    // ---- Lifecycle ----

    fn init(&mut self) -> HalResult<()> {
        if self.initialized {
            return Err(HalError::AlreadyInitialized);
        }

        // Timer — runs from boot, nothing to do.
        let timer = crate::timer::Bcm2710Timer;
        hal::timer::Timer::init(&timer)?;

        // GPIO — no explicit init on BCM2710.

        // Query board information via mailbox.
        self.board_revision = mailbox::get_board_revision().unwrap_or(0);
        self.board_serial   = mailbox::get_board_serial().unwrap_or(0);
        self.board_name     = decode_board_name(self.board_revision);

        self.initialized = true;
        Ok(())
    }

    fn is_initialized(&self) -> bool { self.initialized }

    // ---- Identification ----

    fn info(&self) -> PlatformInfo {
        PlatformInfo {
            platform_id: PlatformId::RpiZero2W,
            arch: Arch::Arm64,
            board_name: self.board_name,
            soc_name: "BCM2710",
            board_revision: self.board_revision,
            serial_number: self.board_serial,
        }
    }

    fn platform_id(&self) -> PlatformId { PlatformId::RpiZero2W }
    fn board_name(&self) -> &'static str { self.board_name }
    fn soc_name(&self) -> &'static str { "BCM2710" }

    // ---- Memory ----

    fn memory_info(&self) -> HalResult<MemoryInfo> {
        let (arm_base, arm_size) = mailbox::get_arm_memory()
            .ok_or(HalError::HardwareFault)?;
        let (gpu_base, gpu_size) = mailbox::get_vc_memory()
            .unwrap_or((0, 0));

        Ok(MemoryInfo {
            arm_base: arm_base as usize,
            arm_size: arm_size as usize,
            peripheral_base: regs::PERIPHERAL_BASE,
            gpu_base: gpu_base as usize,
            gpu_size: gpu_size as usize,
        })
    }

    fn arm_memory(&self) -> usize {
        mailbox::get_arm_memory().map(|(_, s)| s as usize).unwrap_or(0)
    }

    fn total_memory(&self) -> usize {
        let arm = mailbox::get_arm_memory().map(|(_, s)| s as usize).unwrap_or(0);
        let vc  = mailbox::get_vc_memory().map(|(_, s)| s as usize).unwrap_or(0);
        arm + vc
    }

    // ---- Clocks ----

    fn clock_info(&self) -> HalResult<ClockInfo> {
        Ok(ClockInfo {
            arm_freq_hz:  mailbox::get_clock_rate(regs::clock::ARM).unwrap_or(0),
            core_freq_hz: mailbox::get_clock_rate(regs::clock::CORE).unwrap_or(0),
            uart_freq_hz: mailbox::get_clock_rate(regs::clock::UART).unwrap_or(0),
            emmc_freq_hz: mailbox::get_clock_rate(regs::clock::EMMC).unwrap_or(0),
        })
    }

    fn arm_freq_hz(&self) -> u32 {
        mailbox::get_clock_rate(regs::clock::ARM).unwrap_or(0)
    }

    fn arm_freq_measured_hz(&self) -> u32 {
        mailbox::get_clock_measured(regs::clock::ARM).unwrap_or(0)
    }

    fn clock_rate(&self, clock: ClockId) -> u32 {
        hal_to_bcm_clock(clock)
            .and_then(mailbox::get_clock_rate)
            .unwrap_or(0)
    }

    // ---- Temperature ----

    fn temperature_mc(&self) -> HalResult<i32> {
        mailbox::get_temperature()
            .map(|t| t as i32)
            .ok_or(HalError::HardwareFault)
    }

    fn max_temperature_mc(&self) -> HalResult<i32> {
        mailbox::get_max_temperature()
            .map(|t| t as i32)
            .ok_or(HalError::HardwareFault)
    }

    // ---- Throttling ----

    fn throttle_status(&self) -> HalResult<ThrottleFlags> {
        mailbox::get_throttled()
            .map(ThrottleFlags)
            .ok_or(HalError::HardwareFault)
    }

    // ---- Power ----

    fn set_power(&self, device: DeviceId, on: bool) -> HalResult<()> {
        let bcm_id = hal_to_bcm_device(device).ok_or(HalError::InvalidArgument)?;
        if mailbox::set_power_state(bcm_id, on) { Ok(()) } else { Err(HalError::HardwareFault) }
    }

    fn get_power(&self, device: DeviceId) -> HalResult<bool> {
        let bcm_id = hal_to_bcm_device(device).ok_or(HalError::InvalidArgument)?;
        mailbox::get_power_state(bcm_id).ok_or(HalError::HardwareFault)
    }

    // ---- System Control ----

    fn reboot(&self) -> HalResult<()> { Err(HalError::NotSupported) }
    fn shutdown(&self) -> HalResult<()> { Err(HalError::NotSupported) }

    fn panic(&self, _message: &str) -> ! {
        loop { hal::cpu::wfi(); }
    }

    // ---- Debug ----

    fn debug_putc(&self, _c: u8) {
        // TODO: UART output
    }
}
